src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -101,60 +101,54 @@ void AdaptDispatch::_WriteToBuffer(const std::wstring_view string)
             auto measureIt = it;
             while (measureIt && measureIt.GetCellDistance(it) < lineWidth)
             {
-                measureIt++;
+                ++measureIt;
             }
             const auto row = cursorPosition.y;
             const auto cellCount = measureIt.GetCellDistance(it);
             _ScrollRectHorizontally(textBuffer, { cursorPosition.x, row, lineWidth, row + 1 }, cellCount);
         }
-        const auto itEnd = textBuffer.WriteLine(it, cursorPosition, wrapAtEOL, lineWidth - 1);
 
-        if (itEnd.GetInputDistance(it) == 0)
-        {
-            // If we haven't written anything out because there wasn't enough space,
-            // we move the cursor to the end of the line so that it's forced to wrap.
-            cursorPosition.x = lineWidth;
-            // But if wrapping is disabled, we also need to move to the next string
-            // position, otherwise we'll be stuck in this loop forever.
-            if (!wrapAtEOL)
-            {
-                stringPosition++;
-            }
-        }
-        else
+        const auto positionBefore = stringIterator.data();
+        const auto newPosX = textBuffer.Write(cursorPosition.y, cursorPosition.x, lineWidth, wrapAtEOL, attributes, stringIterator);
+        const auto positionAfter = stringIterator.data();
+        const auto failedToWrite = positionBefore == positionAfter;
+        const auto isWrapping = newPosX >= lineWidth;
+
+        if (newPosX != cursorPosition.x)
         {
-            const auto cellCount = itEnd.GetCellDistance(it);
-            const auto changedRect = til::rect{ cursorPosition, til::size{ cellCount, 1 } };
+            const til::rect changedRect{ cursorPosition.x, cursorPosition.y, newPosX, cursorPosition.y + 1 };
             _api.NotifyAccessibilityChange(changedRect);
-
-            stringPosition += itEnd.GetInputDistance(it);
-            cursorPosition.x += cellCount;
         }
 
-        if (cursorPosition.x >= lineWidth)
+        // If we're past the end of the line, we need to clamp the cursor
+        // back into range, and if wrapping is enabled, set the delayed wrap
+        // flag. The wrapping only occurs once another character is output.
+        cursorPosition.x = isWrapping ? lineWidth - 1 : newPosX;
+        cursor.SetPosition(cursorPosition);
+
+        if (isWrapping)
         {
-            // If we're past the end of the line, we need to clamp the cursor
-            // back into range, and if wrapping is enabled, set the delayed wrap
-            // flag. The wrapping only occurs once another character is output.
-            cursorPosition.x = lineWidth - 1;
-            cursor.SetPosition(cursorPosition);
             if (wrapAtEOL)
             {
                 cursor.DelayEOLWrap();
             }
-        }
-        else
-        {
-            cursor.SetPosition(cursorPosition);
+            else if (failedToWrite)
```

Comment:
Are we certain there's no other instances in which this could hit?

Reply:
I would also expect this to happen when we're writing a wide glyph in the last column and autowrap mode _is_ enabled. Because the `TextBuffer::Write` can't handle the wrapping (or if it is, it probably shouldn't be), so it has to fail and let us repeat the loop with the delayed wrap flag set.

When I have a chance I'll try and give this branch a test, but I'm almost certain there's something wrong here.

Reply:
Disregard that - this case is fine. `TextBuffer::Write` still advances the cursor position, so `failedToWrite` is true, but `isWrapping` should also be true, because `newPosX` will have reached the `lineWidth`.

Reply:
`TextBuffer::Write` always returns:
* `colBeg` if the parameters are invalid (= it return the column you gave as the function argument)
* `colExtBeg` otherwise, which is the column past the last character that was written. This includes any padding space that is introduced because a wide character failed to insert!

So in case that we write the last column, and assuming the parameters are valid, this return value will _always_ be equal to `lineWidth`. This is very important, because that's how we detect line wrapping (`isWrapping` is true if `TextBuffer::Write` returns `lineWidth`).

In short, `failedToWrite` can only ever be true if `isWrapping` is true.
...unless I'm mistaken of course.

I'd be extremely happy if you could take a look at this code btw @j4james, if you got the time. Of course this is entirely optional, but the last thing I'd ever want is for this PR to stand in the way of any future plans you have and so if you have any feedback, no matter how critical it is, I'd love to hear it.

Reply:
I'll try and take a look at the code later this weekend, but the `TextBuffer` area is a bit out of my comfort zone, so I probably wouldn't be able to provide a very useful review. That said, I have just rebased all my pending work on top of this branch, and I didn't have any problems with that. So I can at least confirm that this isn't messing with any of my current plans.